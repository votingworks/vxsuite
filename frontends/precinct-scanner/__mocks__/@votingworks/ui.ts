// eslint-disable-next-line import/no-extraneous-dependencies, import/no-import-module-exports
import {
  fakePrintElementWhenReady,
  fakePrintElement,
} from '@votingworks/test-utils';

const ui = jest.requireActual('@votingworks/ui');

module.exports = {
  ...ui,
  printElement: fakePrintElement,
  printElementWhenReady: fakePrintElementWhenReady,
};
