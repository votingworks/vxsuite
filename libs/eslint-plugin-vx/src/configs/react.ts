import recommended = require('./recommended');

const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx'];
const allExtensions = jsExtensions.concat(tsExtensions);

export = {
  ...recommended,
  extends: [
    ...recommended.extends
      .map((name) => (name === 'airbnb-base' ? 'airbnb' : name))
      .filter((name) => name !== 'plugin:prettier/recommended'),
    'airbnb/hooks',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    ...recommended.settings,
    react: {
      version: 'detect',
    },
  },
  rules: {
    ...recommended.rules,
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        controlComponents: ['Select'],
      },
    ],
    'react/jsx-filename-extension': ['error', { extensions: allExtensions }],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-no-bind': 'off',
    'react/jsx-no-constructed-context-values': 'off',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
    'react/require-default-props': 'off',
  },
};
