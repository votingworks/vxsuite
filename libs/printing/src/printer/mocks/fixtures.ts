import {
  IppMarkerInfo,
  PrinterConfig,
  PrinterRichStatus,
  PrinterStatus,
} from '@votingworks/types';

export const MOCK_MARKER_INFO: IppMarkerInfo = {
  color: '#000000',
  highLevel: 100,
  level: 100,
  lowLevel: 2,
  name: 'black cartridge',
  type: 'toner-cartridge',
};

export const MOCK_PRINTER_RICH_STATUS: PrinterRichStatus = {
  state: 'idle',
  stateReasons: [],
  markerInfos: [MOCK_MARKER_INFO],
};

export function getMockConnectedPrinterStatus(
  config: PrinterConfig
): PrinterStatus {
  if (config.supportsIpp) {
    return {
      connected: true,
      config,
      richStatus: MOCK_PRINTER_RICH_STATUS,
    };
  }

  return {
    connected: true,
    config,
  };
}
