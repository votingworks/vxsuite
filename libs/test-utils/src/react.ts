/* eslint-disable no-console -- this file monkey-patches `console` */

/**
 * Suppresses the "Can't perform a React state update on an unmounted component" warning by
 * modifying the `console.error` method to ignore the warning.
 *
 * This is a temporary workaround until we upgrade to React 18.
 */
export function suppressReact17UnmountedWarning(): void {
  const originalConsoleError = console.error;
  console.error = (...args: unknown[]) => {
    if (
      typeof args[0] === 'string' &&
      args[0].startsWith(
        `Warning: Can't perform a React state update on an unmounted component`
      )
    ) {
      return;
    }

    originalConsoleError(...args);
  };
}
