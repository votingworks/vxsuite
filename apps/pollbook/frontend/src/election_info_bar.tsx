import styled from 'styled-components';
import { Caption, Font, LabelledText, DateString, Seal } from '@votingworks/ui';
import type { Election } from '@votingworks/pollbook-backend';

const Bar = styled.div<{ inverse?: boolean }>`
  background: ${(p) => p.inverse && p.theme.colors.inverseBackground};
  color: ${(p) => p.inverse && p.theme.colors.onInverse};
  align-content: flex-end;
  align-items: center;
  border-top: ${(p) => p.theme.sizes.bordersRem.hairline}rem solid
    ${(p) => p.theme.colors.onBackground};
  display: flex;
  gap: 0.5rem;
  justify-content: space-between;
  padding: 0.25rem;
`;

const ElectionInfoContainer = styled.div`
  align-items: center;
  display: flex;
  gap: 0.5rem;
  justify-content: start;
`;

const SystemInfoContainer = styled.div`
  align-content: flex-end;
  flex: 1;
  display: flex;
  flex-wrap: nowrap;
  gap: 0.75rem;
  justify-content: flex-end;
`;

export interface ElectionInfoBarProps {
  election?: Election;
  codeVersion?: string;
  machineId?: string;
  inverse?: boolean;
}
export function ElectionInfoBar({
  election,
  codeVersion,
  machineId,
  inverse,
}: ElectionInfoBarProps): JSX.Element {
  const codeVersionInfo = codeVersion ? (
    <Caption noWrap>
      <LabelledText label="Version">
        <Font weight="bold">{codeVersion}</Font>
      </LabelledText>
    </Caption>
  ) : null;

  const machineIdInfo = machineId ? (
    <Caption noWrap>
      <LabelledText label="Machine ID">
        <Font weight="bold">{machineId}</Font>
      </LabelledText>
    </Caption>
  ) : null;

  if (!election) {
    return (
      <Bar data-testid="electionInfoBar" inverse={inverse}>
        <SystemInfoContainer>
          {codeVersionInfo}
          {machineIdInfo}
        </SystemInfoContainer>
      </Bar>
    );
  }

  const electionInfoLabel = (
    <Font maxLines={2}>
      {election.county.name}, {election.state}
    </Font>
  );

  const electionInfo = (
    <Caption weight="regular">
      <LabelledText labelPosition="bottom" label={electionInfoLabel}>
        <Font weight="bold">{election.title}</Font> â€”{' '}
        <Font noWrap>
          <DateString
            value={election.date.toMidnightDatetimeWithSystemTimezone()}
          />
        </Font>
      </LabelledText>
    </Caption>
  );

  const electionIdInfo = (
    <Caption>
      <LabelledText label="Election ID">
        <Font weight="bold">{election.id}</Font>
      </LabelledText>
    </Caption>
  );

  return (
    <Bar data-testid="electionInfoBar" inverse={inverse}>
      <ElectionInfoContainer>
        <Seal seal={election.seal} maxWidth="2.25rem" inverse={inverse} />
        {electionInfo}
      </ElectionInfoContainer>
      <SystemInfoContainer>
        {codeVersionInfo}
        {machineIdInfo}
        {electionIdInfo}
      </SystemInfoContainer>
    </Bar>
  );
}

const VerticalBar = styled.div<{ inverse?: boolean }>`
  background: ${(p) => p.inverse && p.theme.colors.inverseBackground};
  color: ${(p) => p.inverse && p.theme.colors.onInverse};
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
`;

export function VerticalElectionInfoBar({
  election,
  codeVersion,
  machineId,
  inverse,
}: ElectionInfoBarProps): JSX.Element {
  if (!election) {
    return (
      <VerticalBar inverse={inverse}>
        <Caption>
          {codeVersion && (
            <div>
              Version: <Font weight="semiBold">{codeVersion}</Font>
            </div>
          )}

          {machineId && (
            <div>
              Machine ID: <Font weight="semiBold">{machineId}</Font>
            </div>
          )}
        </Caption>
      </VerticalBar>
    );
  }

  return (
    <VerticalBar inverse={inverse}>
      <ElectionInfoContainer>
        <Seal seal={election.seal} maxWidth="3rem" inverse={inverse} />

        <Caption weight="regular" align="left">
          <Font weight="bold" maxLines={4}>
            {election.title}
          </Font>
          <Font maxLines={4}>
            {election.county.name}, {election.state}
          </Font>
          <div>
            <DateString
              value={election.date.toMidnightDatetimeWithSystemTimezone()}
            />
          </div>
        </Caption>
      </ElectionInfoContainer>

      <Caption>
        {codeVersion && (
          <div>
            Version: <Font weight="semiBold">{codeVersion}</Font>
          </div>
        )}

        {machineId && (
          <div>
            Machine ID: <Font weight="semiBold">{machineId}</Font>
          </div>
        )}

        <div>
          Election ID:
          <Font weight="semiBold">{election.id}</Font>
        </div>
      </Caption>
    </VerticalBar>
  );
}
