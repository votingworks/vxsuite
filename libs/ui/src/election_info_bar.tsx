import React from 'react';

import { ElectionDefinition, PrecinctSelection } from '@votingworks/types';
import { formatShortDate, getPrecinctSelectionName } from '@votingworks/utils';
import styled from 'styled-components';
import { DateTime } from 'luxon';
import { Prose } from './prose';
import { Text, NoWrap } from './text';
import { contrastTheme, Theme } from './themes';
import { Seal } from './seal';

interface BarProps {
  theme?: Theme;
}

const Bar = styled.div<BarProps>`
  display: flex;
  flex-direction: row;
  align-items: end;
  background: ${({ theme: { background } }) => background};
  padding: 0.5rem 0.75rem;
  color: ${({ theme: { color } }) => color};
  gap: 1rem;
`;

export type InfoBarMode = 'voter' | 'pollworker' | 'admin';

interface Props {
  mode?: InfoBarMode;
  electionDefinition: ElectionDefinition;
  codeVersion?: string;
  machineId?: string;
  precinctSelection?: PrecinctSelection;
}
export function ElectionInfoBar({
  mode = 'voter',
  electionDefinition,
  codeVersion,
  machineId,
  precinctSelection,
}: Props): JSX.Element {
  const {
    election: { precincts, date, title, county, state, seal, sealUrl },
  } = electionDefinition;
  const electionDate = formatShortDate(DateTime.fromISO(date));

  /* istanbul ignore next */
  const theme: Theme = { ...(contrastTheme.dark ?? {}) };

  return (
    <Bar theme={theme} data-testid="electionInfoBar">
      {(seal || sealUrl) && <Seal seal={seal} sealUrl={sealUrl} />}
      <Prose maxWidth={false} compact style={{ flex: 1 }}>
        <strong>{title}</strong> â€” <NoWrap>{electionDate}</NoWrap>
        <Text as="div" small>
          {precinctSelection && (
            <React.Fragment>
              <NoWrap>
                {getPrecinctSelectionName(precincts, precinctSelection)},
              </NoWrap>{' '}
            </React.Fragment>
          )}
          <NoWrap>{county.name},</NoWrap> <NoWrap>{state}</NoWrap>
        </Text>
      </Prose>
      {mode !== 'voter' && codeVersion && (
        <Prose maxWidth={false} compact textRight>
          <Text as="div" small noWrap>
            Software Version
          </Text>
          <strong>{codeVersion}</strong>
        </Prose>
      )}
      {mode !== 'voter' && machineId && (
        <Prose maxWidth={false} compact textRight>
          <Text as="div" small noWrap>
            Machine ID
          </Text>
          <strong>{machineId}</strong>
        </Prose>
      )}
      <Prose maxWidth={false} compact textRight>
        <Text as="div" small>
          Election ID
        </Text>
        <strong>{electionDefinition.electionHash.slice(0, 10)}</strong>
      </Prose>
    </Bar>
  );
}
