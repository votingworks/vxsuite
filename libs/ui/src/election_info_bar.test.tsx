import React from 'react';
import { render } from '@testing-library/react';
import { electionSampleDefinition } from '@votingworks/fixtures';
import {
  ALL_PRECINCTS_SELECTION,
  singlePrecinctSelectionFor,
} from '@votingworks/utils';
import { ElectionInfoBar } from './election_info_bar';

test('Renders nothing when there is no election', () => {
  const { container } = render(
    <ElectionInfoBar
      electionDefinition={undefined}
      machineId="0000"
      codeVersion="DEV"
      mode="admin"
    />
  );
  expect(container).toMatchSnapshot();
});

test('Renders ElectionInfoBar without precinct information by default', () => {
  const { container } = render(
    <ElectionInfoBar
      electionDefinition={electionSampleDefinition}
      machineId="0000"
      codeVersion="DEV"
      mode="admin"
    />
  );
  expect(container).toMatchSnapshot();
});

test('Renders ElectionInfoBar with all precincts wording', () => {
  const { container } = render(
    <ElectionInfoBar
      electionDefinition={electionSampleDefinition}
      machineId="0000"
      codeVersion="DEV"
      precinctSelection={ALL_PRECINCTS_SELECTION}
    />
  );
  expect(container).toMatchSnapshot();
});

test('Renders admin ElectionInfoBar with precinct set', () => {
  const { container } = render(
    <ElectionInfoBar
      mode="admin"
      electionDefinition={electionSampleDefinition}
      machineId="0002"
      codeVersion="DEV"
      precinctSelection={singlePrecinctSelectionFor('23')}
    />
  );
  expect(container).toMatchSnapshot();
});
