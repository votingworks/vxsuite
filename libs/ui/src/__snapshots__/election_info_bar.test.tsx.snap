// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ElectionInfoBar with all precincts wording 1`] = `
<div>
  <div
    class="sc-eCImvq dHkUPT"
  >
    <div
      class="sc-bdvvaa fLugxj"
    >
      <strong
        class="sc-dkPtyc gbBfTJ"
      >
        General Election
      </strong>
       —
       
      <span
        class="sc-dkPtyc gbBfTJ"
      >
        Tuesday, November 3, 2020
      </span>
      <div
        class="sc-gsDJrp hIGrZm"
      >
        <span
          class="sc-dkPtyc gbBfTJ"
        >
          All Precincts
        </span>
         
        —
         
        <span
          class="sc-dkPtyc gbBfTJ"
        >
          Franklin County
          ,
           
          State of Hamilton
        </span>
      </div>
    </div>
    <div
      class="sc-jRQAMF kSgACi"
    />
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Election ID
      </div>
      <strong>
        748dc61ad3
      </strong>
    </div>
  </div>
</div>
`;

exports[`Renders ElectionInfoBar without precinct information by default 1`] = `
<div>
  <div
    class="sc-eCImvq dHkUPT"
  >
    <div
      class="sc-bdvvaa fLugxj"
    >
      <strong
        class="sc-dkPtyc gbBfTJ"
      >
        General Election
      </strong>
       —
       
      <span
        class="sc-dkPtyc gbBfTJ"
      >
        Tuesday, November 3, 2020
      </span>
      <div
        class="sc-gsDJrp hIGrZm"
      >
        <span
          class="sc-dkPtyc gbBfTJ"
        >
          Franklin County
          ,
           
          State of Hamilton
        </span>
      </div>
    </div>
    <div
      class="sc-jRQAMF kSgACi"
    />
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Software Version
      </div>
      <strong>
        DEV
      </strong>
    </div>
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Machine ID
      </div>
      <strong>
        0000
      </strong>
    </div>
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Election ID
      </div>
      <strong>
        748dc61ad3
      </strong>
    </div>
  </div>
</div>
`;

exports[`Renders admin ElectionInfoBar with precinct set 1`] = `
<div>
  <div
    class="sc-eCImvq dHkUPT"
  >
    <div
      class="sc-bdvvaa fLugxj"
    >
      <strong
        class="sc-dkPtyc gbBfTJ"
      >
        General Election
      </strong>
       —
       
      <span
        class="sc-dkPtyc gbBfTJ"
      >
        Tuesday, November 3, 2020
      </span>
      <div
        class="sc-gsDJrp hIGrZm"
      >
        <span
          class="sc-dkPtyc gbBfTJ"
        >
          Center Springfield
        </span>
         
        —
         
        <span
          class="sc-dkPtyc gbBfTJ"
        >
          Franklin County
          ,
           
          State of Hamilton
        </span>
      </div>
    </div>
    <div
      class="sc-jRQAMF kSgACi"
    />
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Software Version
      </div>
      <strong>
        DEV
      </strong>
    </div>
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Machine ID
      </div>
      <strong>
        0002
      </strong>
    </div>
    <div
      class="sc-bdvvaa iOQknX"
    >
      <div
        class="sc-gsDJrp hIGrZm"
      >
        Election ID
      </div>
      <strong>
        748dc61ad3
      </strong>
    </div>
  </div>
</div>
`;

exports[`Renders nothing when there is no election 1`] = `<div />`;
